### ISSUE TITLE
Support for Pandas indexes (Cython class `_new_Index`)

### ISSUE THREAD

**Issue by sunwen2000** [jpmml/sklearn2pmml#115]
The error when I exported the pipeline into PMML.
```
In [5]: sklearn2pmml(pipeline, "XGBoost_test_.pmml")
   ...: 




Nov 02, 2018 10:41:09 PM org.jpmml.sklearn.Main run
INFO: Parsing PKL..
Nov 02, 2018 10:41:11 PM org.jpmml.sklearn.Main run
SEVERE: Failed to parse PKL
net.razorvine.pickle.PickleException: expected zero arguments for construction of ClassDict (for pandas.core.indexes.base._new_Index)
	at net.razorvine.pickle.objects.ClassDictConstructor.construct(ClassDictConstructor.java:23)
	at net.razorvine.pickle.Unpickler.load_reduce(Unpickler.java:732)
	at net.razorvine.pickle.Unpickler.dispatch(Unpickler.java:200)
	at org.jpmml.sklearn.PickleUtil$1.dispatch(PickleUtil.java:77)
	at net.razorvine.pickle.Unpickler.load(Unpickler.java:122)
	at org.jpmml.sklearn.PickleUtil.unpickle(PickleUtil.java:98)
	at org.jpmml.sklearn.Main.run(Main.java:103)
	at org.jpmml.sklearn.Main.main(Main.java:93)

Exception in thread "main" net.razorvine.pickle.PickleException: expected zero arguments for construction of ClassDict (for pandas.core.indexes.base._new_Index)
	at net.razorvine.pickle.objects.ClassDictConstructor.construct(ClassDictConstructor.java:23)
	at net.razorvine.pickle.Unpickler.load_reduce(Unpickler.java:732)
	at net.razorvine.pickle.Unpickler.dispatch(Unpickler.java:200)
	at org.jpmml.sklearn.PickleUtil$1.dispatch(PickleUtil.java:77)
	at net.razorvine.pickle.Unpickler.load(Unpickler.java:122)
	at org.jpmml.sklearn.PickleUtil.unpickle(PickleUtil.java:98)
	at org.jpmml.sklearn.Main.run(Main.java:103)
	at org.jpmml.sklearn.Main.main(Main.java:93)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-cf59df20174a> in <module>()
----> 1 sklearn2pmml(pipeline, "XGBoost_Income_EFX_test_2018_11_02.pmml")

/usr/lib/python2.7/site-packages/sklearn2pmml/__init__.pyc in sklearn2pmml(pipeline, pmml, user_classpath, with_repr, debug)
    270                         subprocess.check_call(cmd)
    271                 except CalledProcessError:
--> 272                         raise RuntimeError("The JPMML-SkLearn conversion application has failed. The Java process should have printed more information about the failure into its standard output and/or error streams")
    273         finally:
    274                 if debug:

RuntimeError: The JPMML-SkLearn conversion application has failed. The Java process should have printed more information about the failure into its standard output and/or error streams


```

**Comment by vruusmann** [comment]
Your `pipeline` object contains an instance of CPython class `pandas.core.indexes.base._new_Index`. How did it get there?

Do you have a Python code example about how it happened?

You can immediately solve the problem by reorganizing your pipeline so that the `_new_Index` object would not be included into it.

**Comment by sunwen2000** [comment]
Thanks for reply! Here is the python code for pipeline, basically, I tried to map all features into continuous variables, and run the XGBoost.

```
mapper = DataFrameMapper(
	[([column], [ContinuousDomain()]) for column in features] 
)

regressor = XGBRegressor(
eval_metric="mae", 
early_stopping_rounds=5, 
verbose=True,
learning_rate = 0.1, 
max_depth = 20, 
reg_lambda = 1000, 
reg_alpha=500, 
n_estimators = 100,
objective='reg:linear', 
booster='gbtree', 
n_jobs=-1, 
silent=False)

pipeline = PMMLPipeline([
("mapper", mapper),
("regressor", regressor)
])

pipeline.fit(X_train, y_train) 
sklearn2pmml(pipeline, "XGBoost_Income_EFX_test_2018_11_02.pmml")
```

**Comment by vruusmann** [comment]
Everything seems correct. That leaves `X_train` and `y_train` variables as only suspects.

What are the types of those two variables? `pandas.DataFrame` and `pandas.Series`, respectively, or something else?

You could also try (at least temporarily-) changing them to Numpy arrays. A Numpy array shouldn't be technically be able to contain a `_new_Index` object, because it's a Pandas-only thing.

**Comment by sunwen2000** [comment]
Yes, I tried to make them pandas.DataFrame and pandas.Series. 
```
def readFile(filename):
    df = pd.read_csv(filename,index_col=None)
    df = df.loc[:, ["Y"] + features]
    df = df[(0 < df['Y']) & (df['Y'] < 300000)]
    #df.fillna(-9999, inplace=True)
    return df.loc[:, features], df.loc[:, "Y"]

list_X = []
list_y = []
# Create a pool of processes. By default, one is created for each CPU in your machine.
with concurrent.futures.ProcessPoolExecutor() as executor:
    # Get a list of files to process
    allFiles = glob.glob("/home/xinyu.liu/data/*.csv")
    # Process the list of files, but split the work across the process pool to use all CPUs!
    for f, l in zip(allFiles, executor.map(readFile, allFiles)):
        list_X.append(l[0])
        list_y.append(l[1])

X = pd.concat(list_X)
y = pd.concat(list_y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 123)
```

**Comment by vruusmann** [comment]
Just a guess - maybe it's the `df.loc` syntax that is "contaminating" a regular `pandas.DataFrame` object with `_new_Index` objects?

**Comment by sunwen2000** [comment]
I turn the `X_train` and `y_train` into numpy.array, then turn it back to pandas with column names. Now it works. Thank you very much!

**Comment by vruusmann** [comment]
@sunwen2000 Good to know that this Pandas DataFrame -> Numpy array -> Pandas DataFrame conversion provides a temporary workaround.

However, let us keep this issue open, so that I'll remember to implement a proper fix when I have time.

**Comment by sunwen2000** [comment]
One question that is not relative with this issue: is there any way to make the `sklearn2pmml()` step running in parallel? Moreover, it takes very long time before starting fitting the XGBoost model, I am not sure if it is because of the `mapper` step, or turning the data from pandas into `DMatrix`. Do you have any suggestions to make it faster? Thanks!

**Comment by vruusmann** [comment]
What exactly is the slow operation? Is it `PMMLPipeline.fit(..)` or `sklearn2pmml(..)`?

The former simply calls `sklearn.pipeline.Pipeline.fit(..)`. If that's the expensive operation, then you should open another issue with Scikit-Learn and/or XGBoost projects.

I can only affect `sklearn2pmml(..)`. The `DMatrix` object typically doesn't make it this far.

**Comment by sunwen2000** [comment]
The same error info raises again later even after the `Pandas DataFrame -> Numpy array -> Pandas DataFrame` conversions. It looks that the issue comes from the `pd.concat`, I revised the code into 
`X = pd.concat(list_X, ignore_index=True)
y = pd.concat(list_y, ignore_index=True)`
, and now it works well. Moreover, the `mapper` step above is very slow, I removed them, and now it running much faster.
